### ISSUE TITLE
Should Android workflow host own the props flow and provide a setter for it?

### ISSUE THREAD

**Issue by zach-klippenstein** [square/workflow#1190]
Currently, the `setContentWorkflow` API accepts either a single props value or a stream of props, ie it expects the user to "own" the source of props and updates to props. Internally, our analog to `setContentWorkflow` has a different API: it just provides a setter for the current props value. The host owns the stream of props internally. Framed another way, there are two ways to expose props in the Android runtime host API: require the consumer to declare their props at initialization (state based), or require the consumer to tell the runtime whenever the props change (event based).

I suspect the event based approach is more ergonomic for most apps, since if the root props don't change it doesn't matter, and if they do change then they're probably changing based on events from the OS (new intents, activity results, etc) and maintaining a stream source is just boilerplate for UI wiring code. It's especially inconvenient because `setContentWorkflow` uses a private `ViewModel` to own the runtime but if the consumer needs to maintain a stream, they either need to have their own `ViewModel` or use the `Application` to store the stream source. If the host owned the props stream, consuming code would only need a way to access that view model (probably behind a more tightly scoped interface) to tell it the props changed.

Should we change the API before 1.0?

**Comment by rjrjr** [comment]
I'm not sure how informative our internal stuff is for general use, though. We aren't integrated at the activity level yet.

I think the main use case to consider is intents, and my knee jerk is that they seem props-like. But it's just that, a reflex, haven't tried to do build anything with intent integration yet.

Sounds like a missing demo. Maybe your compose sample could accept intents to change its selection?

**Comment by rjrjr** [comment]
We especially need to keep in mind the tricky timing of intents — how there's already one before onCreate is called,  how it can change at any moment, and how it's sticky. That's what makes it seem like a natural for for props to me.

But again, that's all handwave, haven't tried to live with that choice.

**Comment by zach-klippenstein** [comment]
Agreed, sorry if the description wasn't clear. Assuming intents are a natural and obvious use case for root props, what is the best API to expose for them? 

With what we have now, the most natural thing to write is
```kotlin
class MyActivity {
  override fun onCreate() {
    setContentWorkflow {
      Config(intent.toProps())
    }
  }
}
```

But that's wrong, because it won't get new props from `onNewIntent` and it won't even get new props if the activity is recreated because that lambda is only executed by the view model factory. So instead, the correct thing to do is
```kotlin
class MyActivity {
  // Or similar using another ViewModel
  val props: MutableStateFlow by lazy { application.propsFlow }

  override fun onCreate() {
    props.value = intent.toProps()
    setContentWorkflow {
      Config(props)
    }
  }

  override fun onNewIntent(intent) {
    props.value = intent.toProps()
  }
}
```
But that's a lot of boilerplate to achieve something that is a primary use case. It would be nicer to just be able to do something like
```kotlin
class MyActivity {
  override fun onCreate() {
    setContentWorkflow {
      Config(intent.toProps())
    }
  }

  override fun onNewIntent(intent) {
    // Gets the ViewModel, passes the props to some MutableStateFlow or something that the view model owns.
    setWorkflowProps(intent.toProps())
  }
}
```

**Comment by rjrjr** [comment]
That's a clear improvement.

I wonder if there's a lifecycle method we can hook into to get that more automatically? There are hooks for so many other things that we used to need explicit overrides for. If there were, we could do something like this, where providing a non-null lambda means opting in to intent monitoring:

```kotlin
class MyActivity {
  override fun onCreate() {
    setContentWorkflow {
      Config(::intentToProps)
    }
  }

  private fun intentToProps(intent: Intent): MyPropsType {
    // ...
  }
}
```

**Comment by zach-klippenstein** [comment]
That would be neat, but I haven't seen any framework or AndroidX APIs that would support that.
